add typedefs for src/utils/is.js

ChristopherChudzicki opened the pull request:
Add simple typedefs for most of the functions in `src/utils/is.js`. My immediate motivation was a desire for `isMatrix`, which was missing a type definition. But I just added the typedefs for all the `isWhatever` functions. 

Note: The existing `isThing` typedefs seem to all be for the functions in `src/function/utils`, e.g., `isNumeric`. 

PS: As always, thanks for mathjs. I continue to be amazed at how you maintain this project in such a positive way.

gwhitney commented:
As mentioned in #2431, I think it's putting the cart before the horse a bit to have TypeScript definitions for these functions when they are not even officially published as part of the mathjs API. When we get a response on whether that's a direction mathjs will go, hopefully you can expand this PR to one that makes them full-fledged functions.

That will presumably also involve moving unit tests for these into a location corresponding to their current location in the source (see the TODO at the top of `test/unit-tests/core/typed.test.js`), and if any are missing unit tests (not sure if any are), adding unit tests for those in the correct place.

Is there any reasonable way we can unit test the TypeScript type declarations for mathjs themselves? If so, it would probably be wise to add such tests, as TypeScript issues pop up with fair regularity.

ChristopherChudzicki commented:
@gwhitney Came back to this last night. As discussed in #2444 , the goal of this current PR is to make the type definitions for mathjs more accurately reflect its current exports. To that end, I've added typedefs for all the `src/util/is` isX functions and made Matrix useable as a constructor. 

Glad you opened #2472 else I would not have realized `test:types` exists now, so I added some simple tests for these new types, too.

josdejong commented:
Thanks @ChristopherChudzicki for working this out, I see in the unit tests all the type guards work 😎.

I think this PR is ready to be merged, agree @gwhitney ?

gwhitney commented on types/index.d.ts:
```diff
@@ -2697,7 +2699,86 @@ declare namespace math {
+     isSparseMatrix(x: unknown): x is Matrix;
+
+     isRange(x: unknown): boolean;
+
```
Currently, `math.Range` is exported, and returns a Range object. (I think that's the only way to get one.) In the spirit of matching up the TypeScript typing with the actual JavaScript semantics, the Range type should be declared in this file and this should be changed to a guard for Range. The only question is whether that should be done in this PR, or if it is out of scope for this PR and a separate issue should be filed. Thoughts?

josdejong replied:
Fine with me to address this in a separate PR (to prevent too much scope creep in this PR 😉). But feel free to address right away if you're in the mood for it @ChristopherChudzicki

gwhitney commented on types/index.d.ts:
```diff
@@ -2697,7 +2699,86 @@ declare namespace math {
+     isBoolean(x: unknown): x is boolean;
+
+     isResultSet(x: unknown): boolean;
+
```
This is a similar situation. A ResultSet can be returned from a call to `math.evaluate()`, which at the moment has the less-than-fully specific TypeScript type `any`. In the same spirit, ResultSet should be added to index.d.ts, and the return type of evaluate should be improved, and this should be changed to a guard. So again: in or out of scope for this PR?

ChristopherChudzicki replied:
I'd prefer to add that in a separate PR, which I'm happy to do. The main reason I didn't include it was because it didn't a type declared and I wasn't particularly familiar with it. I've made isChain a type predicate since that already has a type defined.

gwhitney commented on types/index.d.ts:
```diff
@@ -2697,7 +2699,86 @@ declare namespace math {
+     isSymbolNode(x: unknown): x is SymbolNode;
+
+     isChain(x: unknown): boolean;
+
```
Is there a reason `isChain` can't be a guard for MathJsChain? We are defining that type.

gwhitney commented on types/index.d.ts:
```diff
@@ -5128,4 +5213,6 @@ declare namespace math {
   interface ImportObject {
     [key: string]: any;
   }
+
```
Please revert whitespace-only change (insertion of two blank lines) here, unless there's an argument for why adding one blank line is important (but I don't think this codebase uses two blank lines anywhere).

gwhitney commented:
> I think this PR is ready to be merged, agree @gwhitney ?

Ok, so I did a review as I took this to be a request for, just a few comments above. And as far as the big picture goes, let me recap to make sure we are all on the same page: 

* We have decided to keep all of the currently exported "is" functions exported as bona fide mathjs functions.
* Therefore, they need (a) public documentation (not just in-code comments), (b) to be available in the parser, (c) embedded docs, and (d) TypeScript types.
* We have also decided that bona fide mathjs functions, available directly and in the parser, can quite reasonably be ordinary JavaScript functions, not necessarily always typed-functions, and in particular these "is" functions will remain ordinary JavaScript functions, quite possibly "forever."
* Finally, we have decided to split up the "regularization" of these functions into multiple PRs: this one will serve for (d) above, and one or more others will take care of (a)-(c).

If you agree with all those points (I am OK with them) then I think except for the review questions and the fact that I'd suggest that an issue calling for (a)-(c) for the "is" functions be filed before this is merged, I think this is ready to go. I am happy to file the issue when I hear back on this comment; or in fact I think it would be perfectly sensible to morph #2444 into exactly that issue. And @ChristopherChudzicki any chance you're able to try your hand at a PR for some or all of (a)-(c)? It would really help since as you can see there is a lot on the mathjs plate at the moment.

josdejong commented:
Thanks for your clear summary @gwhitney 👍 . Yes please open a new issue to address the things that still need to be addressed after this PR is merged.

gwhitney commented:
On Thu, 2022-03-17 at 15:21 -0700, Chris Chudzicki wrote:
> @ChristopherChudzicki commented on this pull request.
> 
> In types/index.d.ts:
> > +     isMatrix(x: unknown): x is Matrix;
> +
> +     isCollection(x: unknown): x is (Matrix | any[]);
> +
> +     isDenseMatrix(x: unknown): x is Matrix;
> +
> +     isSparseMatrix(x: unknown): x is Matrix;
> +
> +     isRange(x: unknown): boolean;
> +
> +     isIndex(x: unknown): x is Index;
> +
> +     isBoolean(x: unknown): x is boolean;
> +
> +     isResultSet(x: unknown): boolean;
> +
> I'd prefer to add that in a separate PR, which I'm happy to do. The
> main reason I didn't include it was because it didn't a type declared
> and I wasn't particularly familiar with it. I've made isChain a type
> predicate since that already has a type defined.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>

I got this email but don't actually see the comment or the new commit
on GitHub, which seems very weird to me. But in any case, this all
seems reasonable to me and when the commit does show up (don't forget
fixing the whitespace nit as well) I will be happy to merge it and file
an issue with all the outstanding tasks (which yes it would be great if
you could help pitch in on, Christopher).

gwhitney commented:
I see it now and have filed issue #2487 so I will go ahead and merge.

ChristopherChudzicki commented:
@gwhitney Hehe I deleted that comment because I decided to include it as part of an overall comment, so I could collocate the responses about Range and ResultSet. But... I shouldn't have deleted it. I knew that email would be confusing 🤦 

@gwhitney Thanks for the review, I've updated the PR to make `isChain` a type predicate and removed the extra whitespace. I'd prefer to keep larger changes (like defining new Range and ResultSet types) for a subsequent PR. I'm happy to take a stab at that this weekend.

My take-aways from the PR / issue discussions is slightly different. In my opinion:
1. **Types should accurately reflect the current exports of MathJS:** Whether something appears in the markdown documentation or not, if it's exported, it's part of the public API and having types available for exported stuff is useful. (It might not be an *important* part of the public API, in which case removing it in a major release is likely not very disruptive.)

    As an example of this, the `src/util/is` functions are exported, so their types should be accurate.

2. **Runtime type checking for these functions is not worth it:** Since they should *never* throw errors.
3. **Most exported functions should be accessible in the expression parser:** These functions are currently not. But note that some other utility-esque functions are also not available: `create`, `parse`, `evaluate` are three examples.

    Definitely the `isX` functions have a slightly different flavor than `create` and `parse`... `isX` could reasonably be included, whereas I don't see how those other ones could, so I think including them is very reasonable.

In regards to documentation for these `isX` functions... I actually do not think it's worth it, especially if MathJS documentation is handwritten Markdown files (which it appears to be). The `isX` functions are the kind of thing I would include in an auto-generated API reference. Anyone consuming the `isMatrix` function can probably 

IMO, next steps should be:

**Next steps:**
1. Either:
    - Make `isX` etc available in expression parser! Probably not that useful, but
    - OR: remove them (in a major release) from top-level exports, and make them available only via `import { isMatrix } from mathjs/util` or something like that. 
3. Add types for ResultSet and Range
4. Keep improving the types!
5. _Totally unrelated but has been annoying me: Set up linting for the typescript files!_ Or maybe I just have it configured wrong locally.

I'm very happy to help with any and all of those (and elsewhere in MathJs).

ChristopherChudzicki commented:
@gwhitney I should add: Those were my takeaways, but your plan sounds reasonable, too, and I am very happy to contribute to #2487 . The steps in #2487 are very reasonable and useful. Probably better to just do that than dither for a long time, which is my usual mode of operation.

gwhitney commented:
Since I am about to close this, you should probably port your further commentary to #2487. However, @josdejong has confirmed that the desired invariant is: 

mathjs function exports === web-documented functions ==~ functions in parser,

where the last equality is just slightly approximate because of a tiny list of items like "math.create" which don't really make sense in the parser context. But virtually everything else does, including parse and evaluate, which are in fact currently available in the parser.

Also, Jos has officially blessed the "is" functions. Therefore, as recorded in #2487, they do need to be available in the parser, and have JSDoc documentation, which should generate the website documentation, and they should have embedded docs. Help with any/all of these tasks is very welcome!  Thanks for this step :)

josdejong commented:
Published in `v10.4.1`
